src/anml/parameter/parameter.py

```diff
@@ -122,42 +122,159 @@ def design_mat(self, df: pd.DataFrame) -> np.ndarray:
+    `lb` and `ub` are vectors of multiples of ones.
+
+    This type of constraints can be used to impose monotonicity and convexity constraints.
+    For instance, for splines defined on `[0, 5]`, one can specify monotonically decreasing on `[0,1]` with
+    `constr = SplineLinearConstr(x_domain=[0, 1], y_bounds=[-np.inf, 0.0], order=1)`, 
+    monotonically increasing on `[4, 5]` with 
+    `constr = SplineLinearConstr(x_domain=[4, 5], y_bounds=[0.0, np.inf], order=1)`,
+    and overall convexity with
+    `constr = SplineLinearConstr(x_domain=[0, 5], y_bounds=[0.0, np.inf], order=2)`.
+
+    Parameters
+    ----------
+    x_domain: List[float, float]
+        domain for x 
+    y_bounds: List[float, float]
+        bounds for y = Ax
+    order: int
+        order of the derivative
+
+    Raises
+    ------
+    ValueError
+        domain for x is not valid.
+    ValueError
+        bounds for y = Ax is not valid.
+    ValueError
+        invalid derivative order
+    """
+    x_domain: List[float]
+    y_bounds: List[float]
+    order: int
+
+    def __post_init__(self):
+        if self.x_domain[0] >= self.x_domain[1]:
+            raise ValueError('Domain must have positive length.')
+        if self.y_bounds[0] > self.y_bounds[1]:
+            raise ValueError('Lower bound cannot be greater than upper bound.')
+        if self.order < 0:
+            raise ValueError('Order of derivative must be nonnegative.')
+
+
 @dataclass
 class Spline(Variable):
-    """A spline variable.
+    """Spline variable.
+
+    Parameters
+    ----------
+    knots_type : str
+        type of knots. can only be 'frequency' or 'domain'
+    knots_num: int 
+        number of knots 
+    degree: int
+        degree of spines
+    l_linear: bool
+        whether left tail is linear
+    r_linear: bool
+        whether right tail is linear
+    include_intercept: bool
+        whether to include intercept in design matrix 
+    derivative_constr: List[`~anml.parameter.parameter.SplineLinearConstr`]
+        constraints on derivatives 
+    constr_grid_size: int
+        number of points to use when building constraint matrix
+
+    Raises
+    ------
+    VariableError
+        unknown knot type
+    VariableError
+        no covariate has been set
     """
 
     knots_type: str = 'frequency'
     knots_num: int = 3
     degree: int = 3
     l_linear: bool = False
     r_linear: bool = False
+    include_intercept: bool = False
+    derivative_constr: List[SplineLinearConstr] = None
+    constr_grid_size: int = None
 
     def __post_init__(self):
         if self.knots_type not in ['frequency', 'domain']:
             raise VariableError(f"Unknown knots_type for Spline {self.knots_type}.")
         self.num_fe = self.knots_num - self.l_linear - self.r_linear + self.degree - 1
+        if not self.include_intercept:
+            self.num_fe -= 1
 
-    def design_mat(self, df: pd.DataFrame) -> np.ndarray:
+    def create_spline(self, df: pd.DataFrame):
```

Comment:
The covariate extraction requires that each `Variable` has a `design_mat()` call and passes `df` to it, so the signature for all the `design_mat()` have to be the same. Could we change this so that we still have `self.create_spline(df)`, but that's called automatically when we call `self.design_mat(df=df)`, rather than having to be called before?

Reply:
yes, fixed that!